### ISSUE TITLE
Modifying values prop does not always reflect in component

### ISSUE THREAD

**Issue by RookY2K** [sanusart/react-dropdown-select#260]
I've got a use-case that doesn't seem well supported by `react-dropdown-select`(RDS). I am attempting to use RDS as the base for a Tree component, but there seems to be some inconsistencies with how the component handles prop updates under the hood. I've got a `codesandbox` that reproduces the issue that I'll spell out below.

https://codesandbox.io/s/rds-values-prop-inconsistent-fkpr23

In this setup, A1 and A2 are considered children of A. 

Desired interactions:
1) Select A -> A1 and A2 become selected
2) Do (1) then Deselect A -> Deselects A1 and A2 
3) Do (1) then Deselect A1 or A2 -> Deselects A

Actual interactions:
1) Does what is desired
2) Only deselects A and leaves A1 and A2 selected
3) Only deselects A1 or A2 and leaves A and the other child selected

I added some simple console.log statements that show the individual processing steps. I also added a console.warn to show what the actual value of the `values` prop is immediately prior to setting it as a prop for RDS. For both interactions (2) and (3), the `values` prop has the correct value.

**Screenshots**
(2)
![image](https://user-images.githubusercontent.com/12241742/208504079-bde99e61-d473-480b-8a2b-f3c76eec5fa0.png)

![image](https://user-images.githubusercontent.com/12241742/208503859-b3c23c9a-a58b-4f26-9e62-20b07545ea08.png)

(3)
![image](https://user-images.githubusercontent.com/12241742/208504238-1eaafa51-a56b-4ba6-a13e-f3c56fed2b34.png)

![image](https://user-images.githubusercontent.com/12241742/208503651-9a00c4a1-e413-497a-8ace-2e2854f537df.png)

**Comment by sanusart** [comment]
Aah. I can see how this one might become a challenge for RDS. I would assume some internals only consider flat array and not nested with children. I will check options for that.

**Comment by RookY2K** [comment]
> Aah. I can see how this one might become a challenge for RDS. I would assume some internals only consider flat array and not nested with children. I will check options for that.

The "nested-ness" of the options can be considered an implementation detail. The options sent into the RDS component are actually flat. There is just a `children` attribute on any option with children. There is no expectation that RDS knows what to do with that attribute and it should be just ignored.

I overrode the `compareValuesFunc` in the above linked codesandbox just to be sure and the same problems are present. In the override, I made sure that the comparison only cared about the `value` param.

Here is a screenshot of the relevant console output (I added output for the `compareValuesFunc` function:

![image](https://user-images.githubusercontent.com/12241742/208518286-e42881bf-00a5-4870-b76a-f1641e0831b4.png)


TL;DR, the "nested-ness" of the options isn't what is causing this bug.

**Comment by RookY2K** [comment]
Also, just to be doubly certain, I removed the `children` attribute from `A` and just hard-coded some of the utility functions with the correct parent / child relationships. So RDS is _only_ receiving flat options with no nested options. The problem still exists.

**Comment by RookY2K** [comment]
Based on the ordering of the console logs, here is what I _think_ is happening:

1. De-selecting `A` causes two things to happen within the same update cycle:
    1. `A` is removed from internal state leaving `A1` and `A2`
    2. `onChange` is processed by consuming app code and sets the `values` prop to an empty array
2. In `componentDidUpdate` L120, the conditional fails because `prevProps.values`/`this.props.values` are different but so are `prevProps.values` and `prevState.values`

There is an inbuilt assumption that props changes won't happen _right_ after a state change. I.E. The updated values won't be modified in the `onChange` callback. The root of this problem, as far as I can see it, is that I'm modifying the `selectedValues` in the `handleChange` callback and RDS can't handle that. I'm going to put together a simplified sandcodebox that should demonstrate this bug and post a followup comment.

**Comment by RookY2K** [comment]
Here is the simplified codesandbox that more succinctly demonstrates the issue:
https://codesandbox.io/s/rds-onchange-issue-ri5zyo

The basic setup is that in the `handleChange` callback definition, I add the third option to the updated values if it isn't already there. So, the expectation is that whatever option is picked should also cause the third option to be picked. In this example, the third option is `Option D`. 

The screenshot I'm going to post below shows what happens if I choose `Option A`. Option A gets selected, but you can see that the provided `values` prop is `Option A` and `Option D` which is _not_ reflected in the RDS component.
![image](https://user-images.githubusercontent.com/12241742/208526016-4b19c22c-fd47-4dd7-8043-0964d1f42bb8.png)

**RDS internal props / state**
![image](https://user-images.githubusercontent.com/12241742/208526159-abc59c67-386d-42cb-91c4-c0a7671c894a.png)

**Comment by sanusart** [comment]
Isn't it very similar situation to what showed in this example?

https://react-dropdown-select.netlify.app/examples#Custom-Content-And-Dropdown-renderers

Here are several different options selected with one hit. 1st, 2nd and 4th.

**Comment by RookY2K** [comment]
> Here are several different options selected with one hit. 1st, 2nd and 4th

The issue isn't with selecting multiple options at one time. The issue is updating the `values` prop in the `onChange` callback. Any update to `values` in the `onChange` handler will be effectively ignored. 

The cause of the issue is this block in the `componentDidUpdate` method:

```javascript
if (
  !this.props.compareValuesFunc(prevProps.values, this.props.values) &&
  this.props.compareValuesFunc(prevProps.values, prevState.values)
) {
  this.setState(
    {
      values: this.props.values
    },
    () => {
      this.props.onChange(this.state.values);
    }
  );
  this.updateSelectBounds();
}
```

`!this.props.compareValuesFunc(prevProps.values, this.props.values)`

The first predicate in the conditional will evaluate to true as expected since the props changed. 

`this.props.compareValuesFunc(prevProps.values, prevState.values)`

The second predicate, however, evaluates to false because prevState.values reflects the values that triggered the `onChange` whereas prevProps.values reflects the last values that was passed into the RDS component prior to the onChange triggering. 

That means that a valid prop change isn't handled by RDS.

**Comment by sanusart** [comment]
I think it is actually valid. First only compares with props and the second
compares prop with state. I need to recreate in order to see the where is
the described race condition happened.

**Comment by RookY2K** [comment]
To be clear, this _is_ a bug. RDS _should_ be able to handle `onChange` updating the `values` prop. It cannot due to the code block I highlighted above. Here is an example flow:

```javascript
const options = [
  { label: 'Option A', value: 'A' },
  { label: 'Option B', value: 'B' },
  { label: 'Option C', value: 'C' },
  { label: 'Option D', value: 'D' }
];

// Pertinent RDS onComponentUpdate block
if (
  !this.props.compareValuesFunc(prevProps.values, this.props.values) &&
  this.props.compareValuesFunc(prevProps.values, prevState.values)
) {
  this.setState(
    {
      values: this.props.values
    },
    () => {
      this.props.onChange(this.state.values);
    }
  );
  this.updateSelectBounds();
}
```

1. User selected 'Option A'
2. Current state of RDS is updated to look like this: 

```javascript
this.state = {values: [{label: 'Option A', value: 'A'}]
```

3. RDS consumer gets onChange event with updated values being `[{label: 'Option A', value: 'A'}]`
4. RDS consumer updates (via state update) values in `onChange` to be `[{label: 'Option A', value: 'A'}, {label: 'Option C', value: 'C'}]`
5. RDS gets updated props that look like

```javascript
this.props = {values: [{label: 'Option A', value: 'A'}, {label: 'Option C', value: 'C'}]
```

6. In the `onComponentDidUpdate` block I highlighted, here are the values of the pertinent variables:

```javascript
this.props.values = [{label: 'Option A', value: 'A'}, {label: 'Option C', value: 'C'}]
prevProps.values = []
prevState.values = [{label: 'Option A', value: 'A'}]
```

7. When comparing `this.props.values` and `prevProps.values`, they are _not_ equal, so the first predicate is true, as desired.
8. When comparing `this.props.values` and `prevState.values`, they are _also_ not equal, so the second predicate is false, which is _not_ desired.
9. Due to this, RDS state is not updated with `this.props.values` and the passed in `values` prop is effectively ignored.

**Comment by sanusart** [comment]
But this block is getting triggered only on very particular situations. I
dont understand how it is happening. I mean this would never work, why now?

**Comment by sanusart** [comment]
I actually think the bug is on the other block. And the bug is the way props being set - e.g. not in as the callback of setState.

**Comment by RookY2K** [comment]
For the record, I've already developed a workaround to have everything update as expected for my use case. I just wanted to make sure that the bug was brought to your attention.
My naive approach to fixing this bug would be to remove the second predicate:
`this.props.compareValuesFunc(prevProps.values, prevState.values)`

But I haven't tested that, and I'm sure there are multiple reasons why that would be a bad idea. This would _probably_ be easier to handle in a FunctionComponents using `useEffect` hooks rather than relying on the `componentDidUpdate`, but that seems like a pretty large rewrite.

**Comment by RookY2K** [comment]
Also, to be clear, this is _not_ an internal react issue. Due to how RDS works, clicking an option updates `values` in state. It isn't until that state update has completed that the `onChange` event is triggered. Therefore, when the next prop update occurs as a response to the `onChange`, `prevState` will now be equivalent to `this.state` in RDS.

**Comment by sanusart** [comment]
> 

Functional component rewrite indeed will be a huge change. But this is something that will need to happened. Eventually they will stop supporting lifecycles I guess.

**Comment by sanusart** [comment]
> I just wanted to make sure that the bug was brought to your attention

And I appreciate it.

**Comment by RookY2K** [comment]
> Bug on RDS or not, I never said it was an issue on react, I think it is prevention of loop. I might be wrong. But updating state and triggering update on the same time sounds like something react would try to prevent and optimize.

My apologies, I didn't mean to put words into your mouth. 

This is getting to the level of pedantic, but I just wanted it to be clear that this wasn't react optimizing or preventing loops. In fact, it seems like it is actually RDS that is attempting to prevent a loop. React _does_ have heuristics to optimize _how_ and _when_ state is updated, but none (that I know of) to prevent a loop. In any event, the bug happens _after_ state is updated in RDS. 

In any event, you are on the case, so I'll let you do your thing. Thank you for looking into this.

**Comment by sanusart** [comment]
All is good 😊
